components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/RuleManagementDAOImpl.java

```diff
@@ -0,0 +1,282 @@
+/*
+ * Copyright (c) 2024, WSO2 LLC. (http://www.wso2.com).
+ *
+ * WSO2 LLC. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.rule.management.dao.impl;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.wso2.carbon.database.utils.jdbc.NamedJdbcTemplate;
+import org.wso2.carbon.database.utils.jdbc.exceptions.DataAccessException;
+import org.wso2.carbon.database.utils.jdbc.exceptions.TransactionException;
+import org.wso2.carbon.identity.core.util.IdentityDatabaseUtil;
+import org.wso2.carbon.identity.rule.management.constant.RuleSQLConstants;
+import org.wso2.carbon.identity.rule.management.dao.RuleManagementDAO;
+import org.wso2.carbon.identity.rule.management.exception.RuleManagementException;
+import org.wso2.carbon.identity.rule.management.exception.RuleManagementServerException;
+import org.wso2.carbon.identity.rule.management.model.Expression;
+import org.wso2.carbon.identity.rule.management.model.Rule;
+import org.wso2.carbon.identity.rule.management.model.Value;
+import org.wso2.carbon.identity.rule.management.model.internal.ANDCombinedRule;
+import org.wso2.carbon.identity.rule.management.model.internal.ORCombinedRule;
+
+import java.io.ByteArrayInputStream;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+
+/**
+ * Rule Management DAO Implementation.
+ * This class is used to perform CRUD operations on Rule in the database.
+ */
+public class RuleManagementDAOImpl implements RuleManagementDAO {
+
+    /**
+     * This method will add the Rule to the database and add the Rule Value References to the database.
+     *
+     * @param rule     Rule object
+     * @param tenantId Tenant ID
+     * @throws RuleManagementException If an error occurs while adding the rule to the database.
+     */
+    @Override
+    public void addRule(Rule rule, int tenantId) throws RuleManagementException {
+
+        try {
+            addRuleToDB(rule, tenantId);
```

Comment:
If we want to following tasks in one transaction, then we need to initiate the jdbcTemplate.withTransaction() from this level

Reply:
There seems to be some issue with the current utility that handles transactions. When data is not committed if the rest of the updates are followed on top of uncommitted data they seem to fail.
https://github.com/wso2-enterprise/asgardeo-product/issues/27910

We have to look at this separately. So for now for rule addition transaction flow will be avoided. More over, looks like we haven't seem to have tried this transactional flow in similar scenarios anywhere in the product. I couldn't locate any even though I looked for references

Reply:
As discussed adding 'withTransaction' for every query started using the same connection and resolved this issue. Thus this is updated accordingly